File: python_impl/src/controllers/controllers/example_traj.py
import rclpy
import numpy as np
from rclpy.node import Node

from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint


class ExampleTraj(Node):

    def __init__(self):
        super().__init__('minimal_publisher')

        self._HOME = [np.deg2rad(0), np.deg2rad(40),
                     np.deg2rad(30), np.deg2rad(-30)]
        self._beginning = self.get_clock().now()
        self._publisher = self.create_publisher(JointTrajectory, 'joint_cmds', 10)
        timer_period = 0.04  # seconds
        self._timer = self.create_timer(timer_period, self.timer_callback)

    def timer_callback(self):
        now = self.get_clock().now()
        msg = JointTrajectory()
        msg.header.stamp = now.to_msg()

        dt = (now - self._beginning).nanoseconds * (1e-9)
        
        point = JointTrajectoryPoint()
        point.positions = [self._HOME[0] + 0.1 * np.pi * np.sin(2 * np.pi / 10.0 * dt),
                           self._HOME[1] + 0.1 * np.pi * np.sin(2 * np.pi / 10.0 * dt),
                           self._HOME[2] + 0.25 * np.pi * (np.sin(2 * np.pi / 10.0 * dt) - 1),
                           self._HOME[3] + 0.25 * np.pi * np.sin(2 * np.pi / 10.0 * dt),
                           0.5 * np.sin(2 * np.pi / 10.0 * dt) + 0.5
                           ]
        msg.points = [point]

        self._publisher.publish(msg)
        

def main(args=None):
    rclpy.init(args=args)

    example_traj = ExampleTraj()

    rclpy.spin(example_traj)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    example_traj.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp_impl/src/controllers/src/example_traj.cpp
#include "example_traj.hpp"

constexpr float DEG2RAD = M_PI / 180.0;

ExampleTraj::ExampleTraj() :
  rclcpp::Node("example_traj"),
  HOME{DEG2RAD * 0, DEG2RAD * 40, DEG2RAD * 30, DEG2RAD * -30}
{
    using namespace std::chrono_literals;

    this->_beginning = this->now();
    
    this->_publisher = this->create_publisher<trajectory_msgs::msg::JointTrajectory>("joint_cmds", 10);
    this->_timer = this->create_wall_timer(
      75ms, std::bind(&ExampleTraj::_timer_callback, this));
}

void ExampleTraj::_timer_callback()
{
  auto now = this->now();
  auto msg = trajectory_msgs::msg::JointTrajectory();
  msg.header.stamp = now;
  
  double dt = (now - this->_beginning).seconds();
  auto point = trajectory_msgs::msg::JointTrajectoryPoint();
  point.positions = {HOME[0] + 0.1 * M_PI * sin(2.0 * M_PI / 10.0 * dt),
                    HOME[1] + 0.1 * M_PI * sin(2.0 * M_PI / 10.0 * dt),
                    HOME[2] + 0.1 * M_PI * sin(2.0 * M_PI / 10.0 * dt),
                    HOME[3] + 0.1 * M_PI * sin(2.0 * M_PI / 10.0 * dt),
                    0.5 * sin(2 * M_PI / 10.0 * dt) + 0.5
                  };

  msg.points = {point};

  this->_publisher->publish(msg);

}

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
    
  rclcpp::spin(std::make_shared<ExampleTraj>());
  rclcpp::shutdown();
  return 0;
}
